dpdk/vr_dpdk_virtio.c

```diff
@@ -108,10 +139,45 @@ vr_dpdk_virtio_rx_queue_init(unsigned int lcore_id, struct vr_interface *vif,
     rx_queue->rxq_burst_size = VR_DPDK_VIRTIO_RX_BURST_SZ;
     rx_queue->q_vif = vif;
 
+    /* store queue params */
+    rx_queue_params->qp_release_op = &dpdk_virtio_rx_queue_release;
+
     return rx_queue;
 }
 
 /*
+ * dpdk_virtio_tx_queue_release - releases a virtio TX queue.
+ *
+ * Returns nothing.
+ */
+static void
+dpdk_virtio_tx_queue_release(unsigned lcore_id, struct vr_interface *vif)
+{
+    struct vr_dpdk_lcore *lcore = vr_dpdk.lcores[lcore_id];
+    struct vr_dpdk_queue *tx_queue = &lcore->lcore_tx_queues[vif->vif_idx];
+    struct vr_dpdk_queue_params *tx_queue_params
+                        = &lcore->lcore_tx_queue_params[vif->vif_idx];
+
+    tx_queue->txq_ops.f_tx = NULL;
+    rte_wmb();
+
+    /* remove the ring from the list of rings to push */
+    dpdk_ring_to_push_remove(tx_queue_params->qp_ring.host_lcore_id,
```

Comment:
When is dpdk_ring_to_push_add() called on the virtio TX queue? The virtio interface has 1 queue, so one lcore will call  vr_dpdk_virtio_tx_queue_init(). The other lcores will call vr_dpdk_ring_tx_queue_init(). vr_dpdk_ring_tx_queue_init() calls dpdk_ring_to_push_add(), but that queue already has a qp_release_op to call dpdk_ring_tx_queue_release(). I am not sure when dpdk_ring_to_push_remove() is needed above.